Oracle crawler (#931)
MetaphorData/connectors: metaphor/oracle/README.md

# Oracle Connector

> This connector extracts technical metadata from a MySQL database using [oracledb](https://oracle.github.io/python-oracledb/) library.

Reviewer: Oracle
```suggestion
This connector extracts technical metadata from an Oracle database using [oracledb](https://oracle.github.io/python-oracledb/) library.
```

## Setup

> We recommend you to create a role

Reviewer: Complete the sentence?

```sql
-- Step 1: Create the role
CREATE ROLE metaphor;

-- Grant necessary privileges to the role
GRANT CREATE SESSION TO metaphor; -- Allows login
GRANT SELECT_CATALOG_ROLE TO metaphor; -- Allows get DDL

-- Step 2: Create the user
CREATE USER metaphor_user IDENTIFIED BY my_password;

-- Step 3: Assign the role to the user
GRANT metaphor TO metaphor_user;

```

> We also need `SELECT` privileges on table and views to get metadata

Reviewer: Can doc we can point to?

Author: https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/ALL_TABLES.html

> ALL_TABLES describes the relational tables accessible to the current user. To gather statistics for this view, use the DBMS_STATS package.

Not sure if "accessible" means select

Author: But I can't see tables belongs to a schema if I don't grant `SELECT` to crawler user

```sql
BEGIN
    FOR t IN (SELECT table_name, owner
              FROM all_tables 
              WHERE owner = 'TARGET_SCHEMA') 
    LOOP
        EXECUTE IMMEDIATE 'GRANT SELECT ON ' || t.owner || '.' || t.table_name || ' TO metaphor';
    END LOOP;
END;
```

## Config File

Create a YAML config file based on the following template.

### Required Configurations

If using user password authentication:

```yaml
host: <hostname>  # the host the crawler will connect to
user: <username>
password: <password>
database: <default_database_for_connections>
```

### Optional Configurations

> By default, the connector will connect using the default MySQL port 3306. You can change it using the following config:

Reviewer: ```suggestion
By default, the connector will connect using the default Oracle port 1521. You can change it using the following config:
```

```yaml
port: <port_number>
```

> If you are indirectly connect to the database, you need to set the real host of the database using the following config:

Reviewer: Not sure I understand the use of this config?

Author: Like account to overwrite. I think we have some similar issues, users may run the crawler on different network environment so the `alternative_host` can be specified to host won't change a log.

Author: Updated the doc
